### PR TITLE
[SPARK-8501] [SQL] Avoids reading schema from empty ORC files (backport to 1.4)

### PR DESCRIPTION
**Pull request by liancheng**
This PR backports #7199 to branch-1.4

### CODE REVIEW

**Comment by liancheng** [comment]
cc @yhuai @zhzhan

**Comment by zhzhan** [comment]
@liancheng Because in spark, we will not create the orc file if the record is empty. It is only happens with the ORC file created by hive, right?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileOperator.scala [hunk 1]
```diff
@@ -24,30 +24,58 @@ import org.apache.hadoop.hive.serde2.objectinspector.StructObjectInspector
...
+    def isWithNonEmptySchema(path: Path, reader: Reader): Boolean = {
+      reader.getObjectInspector match {
+        case oi: StructObjectInspector if oi.getAllStructFieldRefs.size() > 0 =>
+          true
+        case oi: StructObjectInspector if oi.getAllStructFieldRefs.size() == 0 =>
+          logInfo(
+            s"ORC file $path has empty schema, it probably contains no rows. " +
+              "Trying to read another ORC file to figure out the schema.")
+          false
+        case _ => false
```
Comment (zhzhan) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/orc/OrcFileOperator.scala]: In what situation, will the third case happen? If not exist, can we collapse the 2nd and 3rd case?

**Comment by zhzhan** [comment]
some minor comments. Overall, LGTM

**Comment by liancheng** [comment]
@zhzhan Thanks for the review! Updated.

Actually while writing Javadoc of `OrcFileOperator.getFileReader`, I feel the semantics of this method is kinda weird and need some refactoring. But let's leave this to a follow-up PR since 1.4.1 RC2 is being cut soon and I'd like to include this one in 1.4.1.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala [hunk 2]
```diff
@@ -44,7 +44,7 @@ abstract class OrcSuite extends QueryTest with BeforeAndAfterAll {
     import org.apache.spark.sql.hive.test.TestHive.implicits._
 
     sparkContext
-      .makeRDD(1 to 100)
+      .makeRDD(1 to 10)
```
Comment (liancheng) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: The numbers were increased to 100 to workaround SPARK-8501. Now it's fixed, so revert them back.

**Comment by liancheng** [comment]
Merging to branch-1.4.

**Comment by sarutak** [comment]
This PR has already merged right?
It's funny that this PR is still open.
@liancheng Mind manually closing?

**Comment by liancheng** [comment]
@sarutak Thanks for reminding, closing it :)

**Pull request closed without merging**